openzwave & pythonPackages.python_openzwave: init at 20180404 and 0.4.4

etu opened the pull request:
###### Motivation for this change
OpenZWave can be used with a wide range of USB-Sticks used to controll Z-Wave devices.

Python_OpenZWave can be used with homeassistant that is already packaged to be able to control and script Z-Wave devices.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

etu committed:
openzwave: Init at 2018-04-04

etu committed:
python36Packages.python_openzwave: init at 0.4.4

etu commented:
cc @dotlambda @f-breidenstein maybe?

dotlambda commented on pkgs/misc/drivers/openzwave/default.nix:
```diff
@@ -0,0 +1,51 @@
+, udev }:
+
+let
+  version = "2018-04-04";
```
Why not a stable version?

etu replied:
Because the latest stable is V1.5 and was relaseaed 2016-08-15. And the repository seems to be quite active since then.

https://github.com/OpenZWave/open-zwave/releases/tag/V1.5

dotlambda replied:
Seems like almost all distros package 1.5: https://repology.org/metapackage/openzwave/versions.
But `pip install python_openzwave` installs from master by default: https://github.com/OpenZWave/python-openzwave#python-openzwave-04x-is-here-.
So I guess using an unstable version should be fine.
You could however ask them what they want distributions to do on their mailing list.

dotlambda commented on pkgs/misc/drivers/openzwave/default.nix:
```diff
@@ -0,0 +1,51 @@
+    makeFlagsArray+=("DESTDIR=$out")
+    makeFlagsArray+=("PREFIX=$out")
+    makeFlagsArray+=("pkgconfigdir=lib/pkgconfig")
+    makeFlagsArray+=("sysconfigdir=etc/openzwave")
```
You can set this in `makeFlags`, no need to use `preBuild`. See https://nixos.org/nixpkgs/manual/#build-phase.

etu replied:
That is much nicer, thanks :+1:

etu replied:
It doesn't seem to work like the docs suggests:

```
  makeFlags = [
    "DESTDIR=$(out)"
    "PREFIX=$(out)"
    "pkgconfigdir=lib/pkgconfig"
    "sysconfigdir=etc/openzwave"
  ];
```

Doesn't build and it seems like the issue is with `$out` because it outputs this:
```
build flags: SHELL=/nix/store/q1g0rl8zfmz7r371fp5p42p4acmv297d-bash-4.4-p19/bin/bash DESTDIR=\$\(out\) PREFIX=\$\(out\) pkgconfigdir=lib/pkgconfig sysconfigdir=etc/openzwave
```

etu replied:
Doesn't help at all :D

dotlambda replied:
I tried myself and it gives me a different error when leaving out parantheses: `mv: cannot stat '/nix/store/nhy5v0l6yccgi2wfp01d457dp9g6ql9w-openzwave-2018-04-04//nix/store/nhy5v0l6yccgi2wfp01d457dp9g6ql9w-openzwave-2018-04-04/bin': No such file or directory`.

etu replied:
Yeah, that's the fixupPhase

dotlambda commented on pkgs/misc/drivers/openzwave/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+  preBuild = ''
+    makeFlagsArray+=("DESTDIR=$out")
+    makeFlagsArray+=("PREFIX=$out")
```
Try removing this line and you shouldn't need the fixup anymore.

etu replied:
Removing that created a whole range of other issues, because there's a thing that doesn't respect the `DESTDIR`...

So currently it creates:
```
/nix/store/xyz-openzwave-2018-04-04/lib/pkgconfig/libopenzwave.pc
/nix/store/xyz-openzwave-2018-04-04/nix/store/xyz-openzwave-2018-04-04/<rest-of-the-files>
```

I haven't found a good workaround for that :/

etu replied:
Ok, I've gone through all the things now.

I've managed to set `PREFIX=` (as in empty string to not get the default `/usr/local/bin`).

And got rid of  most of fixup but had to patch the pkgconfig-file so the python library can be built.

I have tested it with the python library on x86_64 with home assistant as well. So at the moment it works and is slightly less ugly :)

dotlambda commented on pkgs/misc/drivers/openzwave/default.nix:
```diff
@@ -0,0 +1,51 @@
+
+  meta = with stdenv.lib; {
+    description = "C++ library to control Z-Wave Networks via a USB Z-Wave Controller";
+    homepage = https://github.com/OpenZWave/open-zwave;
```
Maybe http://www.openzwave.net/ instead?

etu replied:
I've honestly missed that that one exists because it's not coming up top when I've been doing searches :D

dotlambda reviewed (commented):
Can you explain what `/etc/openzwave` usually contains?

Did you test that this actually works with Home Assistant?

etu commented:
@dotlambda `/etc/openzwave` contains a bunch of XML files that are configuration for OpenZWave. It's needed to be able to use OpenZWave at all. All utilities and libs in all languages depends on it.

etu commented:
@dotlambda As for your other question:
> Did you test that this actually works with Home Assistant?

Yes, I have it running here on my desktop as well as a raspberry pi. I had to change one little thing in the home-assistant derivation (for some depenedency), and had to add some configuration myself to the config file.

Diff of hame-assistant:
```diff
diff --git a/pkgs/servers/home-assistant/default.nix b/pkgs/servers/home-assistant/default.nix
index 75f8b309a40..e1814317463 100644
--- a/pkgs/servers/home-assistant/default.nix
+++ b/pkgs/servers/home-assistant/default.nix
@@ -72,6 +72,7 @@ in with py.pkgs; buildPythonApplication rec {
   propagatedBuildInputs = [
     # From setup.py
     requests pyyaml pytz pip jinja2 voluptuous typing aiohttp async-timeout astral certifi attrs
+    pydispatcher
     # From http, frontend and recorder components
     sqlalchemy aiohttp-cors hass-frontend
   ] ++ componentBuildInputs ++ extraBuildInputs;
```

configuration:
```nix
  # Enable Home Assistant, open port and add the hass user to the dialout group
  services.home-assistant.enable = true;
  services.home-assistant.package = pkgs.home-assistant.override {
    extraPackages = ps: [ pkgs.python36Packages.python_openzwave ];
  };

  networking.firewall.allowedTCPPorts = [ 8123 ];
  users.extraUsers.hass.extraGroups = [ "dialout" ];
```

Yaml for home-assistant:
```yml
zwave:
  usb_path: /dev/serial/by-id/usb-0658_0200-if00
  config_path: /nix/store/dlgdch7iwdd6ky4z1b4wpzrn1mg8w4i5-openzwave-2018-04-04/etc/openzwave/
```

etu commented on pkgs/misc/drivers/openzwave/default.nix:
```diff
@@ -0,0 +1,44 @@
+    makeFlagsArray+=("PREFIX=")
+    makeFlagsArray+=("pkgconfigdir=lib/pkgconfig")
+    makeFlagsArray+=("sysconfigdir=etc/openzwave")
+  '';
```
@dotlambda At least I've got "rid" of the prefix and lots of file moving in the fixup, but still not sure how to use `makeFlags`... At least it's quite alot cleaner overall now....

dotlambda commented:
> `extraPackages = ps: [ pkgs.python36Packages.python_openzwave ];`

That shoud be `extraPackages = ps: with ps; [ python_openzwave pydispatcher ];`.
However, after running `pkgs/servers/home-assistant/parse-requirements.py`, these dependencies will be added automatically when using the `services.home-assistant.config` option or `extraComponents = [ "zwave" ]`.

dotlambda commented:
You need to copy exactly that: `extraPackages = ps: with ps; [ python_openzwave pydispatcher ];`

etu commented:
Oh, yeah, that did work :)

dotlambda commented on pkgs/top-level/all-packages.nix:
```diff
@@ -20532,6 +20532,8 @@ with pkgs;
 
   moltengamepad = callPackage ../misc/drivers/moltengamepad { };
 
+  openzwave = callPackage ../misc/drivers/openzwave { };
```
I think this belongs in `pkgs/development/libraries`.

dotlambda commented:
How about so?
```
  installFlags = [
    "PREFIX="
    "pkgconfigdir=lib/pkgconfig"
    "sysconfigdir=etc/openzwave"
  ];

  preInstall = ''
    installFlagsArray+=("DESTDIR=$out")
  '';
```

@etu Please also try fixing the `Fontconfig error: Cannot load default config file` errors.

Maybe @peterhoeg or @FRidh want to have a look at this? I'm certainly not an expert when it comes to C libraries.

Commit:
home-assistant: treat "-" and "_" equally in parse-requirements.py

dotlambda commented:
@etu I added a commit which should fix `extraComponents = [ "zwave" ];`.

FRidh commented on pkgs/servers/home-assistant/parse-requirements.py:
```diff
@@ -64,9 +63,11 @@ def name_to_attr_path(req):
         names.append(req[len('python-'):])
     for name in names:
+        # treat "-" and "_" equally
+        name = re.sub('[-_]', '[-_]', name)
```
replacement is a regex as well?

dotlambda replied:
Yeah, because I need to replace both, - and _ at the same time. I don't think `string.replace()` can do that.

dotlambda replied:
I could do
```py
name = name.replace('_', '-')
name = name.replace('-', '[-_]')
```

etu commented:
@dotlambda Re fontconfig: Here I have a build log of openzwave and python_openzwave: https://gist.github.com/etu/6bdba2f432a839566e6806b5d72e3773 -- which warnings do you refer to?

etu commented:
@dotlambda 
I've copied the files of that commit to my own `nixpkgs` and tried to change the config to:
```
  # Enable Home Assistant, open port and add the hass user to the dialout group
  services.home-assistant.enable = true;
  services.home-assistant.package = pkgs.home-assistant.override {
    extraComponents = [ "zwave" ];
  };

  networking.firewall.allowedTCPPorts = [ 8123 ];
  users.extraUsers.hass.extraGroups = [ "dialout" ];
```

That resulted in a python error log from home-assistant that complaints about a missing python module named `openzwave`. So I'm not sure I trust `extraComponents` at all :p

dotlambda commented:
Maybe that error only appears when enabling sandboxing.
[log.txt](https://github.com/NixOS/nixpkgs/files/1887268/log.txt)

Commit:
home-assistant: treat extraComponents with multiple requirements correctly

etu commented:
@dotlambda 

This patch does get rid of the fontconfig warnings, but I'm not sure where to put the environment variables because preInstall feels like the wrong place.

```diff
diff --git a/pkgs/development/libraries/openzwave/default.nix b/pkgs/development/libraries/openzwave/default.nix
index 0d5feae2419..3c18849f1c5 100644
--- a/pkgs/development/libraries/openzwave/default.nix
+++ b/pkgs/development/libraries/openzwave/default.nix
@@ -1,5 +1,5 @@
 { stdenv, fetchFromGitHub
-, doxygen, libxml2, git, which, pkgconfig, graphviz-nox
+, doxygen, libxml2, git, which, pkgconfig, graphviz-nox, fontconfig
 , udev }:
 
 let
@@ -15,7 +15,7 @@ in stdenv.mkDerivation rec {
     sha256 = "0yby8ygzjn5zp5vhysxaadbzysqanwd2zakz379299qs454pr2h9";
   };
 
-  nativeBuildInputs = [ doxygen libxml2 git which pkgconfig graphviz-nox ];
+  nativeBuildInputs = [ doxygen libxml2 git which pkgconfig graphviz-nox fontconfig ];
   buildInputs = [ udev ];
 
   installFlags = [
@@ -26,6 +26,9 @@ in stdenv.mkDerivation rec {
 
   preInstall = ''
     installFlagsArray+=("DESTDIR=$out")
+
+    export FONTCONFIG_FILE=${fontconfig.out}/etc/fonts/fonts.conf
+    export FONTCONFIG_PATH=${fontconfig.out}/etc/fonts/
   '';
 
   postPatch = ''
```

I will try your change.

dotlambda commented:
You can make these a part of the Nix expression, i.e.
```
preInstall = ''
  installFlagsArray+=("DESTDIR=$out")'
'';

FONTCONFIG_FILE = "${fontconfig.out}/etc/fonts/fonts.conf";
FONTCONFIG_PATH = "${fontconfig.out}/etc/fonts/";
```

etu commented:
@dotlambda Oh, ok. Rebased and pushed.

I've tested that change in home-assistant/default.nix and no such luck. Now it's missing `pydispatcher` instead :confused:

etu commented:
@dotlambda I've just confirmed that my patch here: https://github.com/NixOS/nixpkgs/pull/38577/files#diff-588a05cff755242c814f80436a1184faR23

Works properly in home-assistant. So one doesn't need to configure that. That makes everything better I'd say :smile:

peterhoeg commented:
Sorry about being late to the party. Please don't merge this yet, I have a few comments.

etu commented:
@peterhoeg Oh, nice. I'm gonna test through it all tonight. You can push this to my branch if you want :)

peterhoeg commented:
Pushed. You probably want to squash those commits into the relevant commits on your side.

etu commented:
@peterhoeg git is used by openzwave build to find version-number from it's git-repo. So without it I got a warning that it doesn't exist. I don't think it matters very much for the actual usage though.

peterhoeg commented:
Correct, you get the warning but $src is not a checked out git repo but simply a directory, so it cannot do anything there anyway.

etu commented:
@peterhoeg I've rebased and squashed your changed. And I'll fixed some dependency issues for the CLI tools for the python library and have pushed it. I hope it looks good now :)

peterhoeg commented:
Looks good. @dotlambda/@fridh do you want this to go via staging?

dotlambda commented:
I manually checked that it won't create merge conflict, so I guess merging into master is fine.

dotlambda commented:
Thanks a lot @etu!
